net/dns/manager.go

```diff
@@ -339,6 +341,9 @@ func (m *Manager) Query(ctx context.Context, bs []byte, from netaddr.IPPort) ([]
 		// continue
 	}
 
+	ctx, cancel := context.WithTimeout(ctx, dnsQueryTimeout)
```

Comment:
This can be moved to after the maxActiveQueries right before the `m.resolver.Query` call.

But why isn't it in `(*Resolver).Query`, actually?

Or even inside `(*Resolver).Query` it only needs to be done in the `errNotOurName` forwarding path, otherwise we'll answer right away (or have an answer right away), meaning you can avoid the context creation work.

In fact, we're already creating a new context in `(*forwarder).forwardWithDestChan`:

```go
	// Guarantee that the ctx we use below is done when this function returns.
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()
```

That `WithCancel` could be a `WithTimeout` and not increase the total number of context wrappers created.

But lawlz... `(*Resolver).Query` also has a `ctx, cancel := context.WithCancel(ctx)` with a deferred cancel.

Everybody loves that pattern so every layer does it I guess.

In retrospect I think the middle layer might be best: as you want to nuke `forwarderWithDestChan` anyway.

So move it to the immediate caller of forwarding, the `ctx, cancel := context.WithCancel(ctx)` in `(*Resolver).Query`, changing that to a `WithTimeout`?

Reply:
Done